### PR TITLE
Show users a warning when they are sending directly to a token contract

### LINKED ISSUE #12642: Warning not being shown when sending tokens to a known token contract address
In `ducks/send/send.js` we have code that is meant to show a warning to users when they are sending to a known contract address:

```
if (
          isSendingToken &&
          isValidHexAddress(recipient.userInput) &&
          (tokenAddressList.find((address) =>
            isEqualCaseInsensitive(address, recipient.userInput),
          ) ||
            checkExistingAddresses(recipient.userInput, tokens))
        ) {
          recipient.warning = KNOWN_RECIPIENT_ADDRESS_WARNING;
        } else {
          recipient.warning = null;
 }
```

This `recipient.warning` property is read in `renderDialogs` of `ui/pages/send/send-content/add-recipient/add-recipient.component.js`

Currently, if I try to send DAI token to the DAI contract, no warning is shown.

Acceptance criteria:
- if a known token contract address is entered in the recipient field in the send flow, a warning should be shown to the user
- this warning can just be the `<Dialog type="warning" className="send__error-dialog">` that `add-recipient.component.js` is currently meant to render
- "known token contract addresses" include any of the addresses in this list: https://github.com/MetaMask/contract-metadata/blob/master/contract-map.json

### LINKED ISSUE #13486: Improve "Known Contract Address" warning
### Describe the bug

Background: See [My reddit post](https://www.reddit.com/r/CryptoCurrency/comments/sgdxtg/psa_we_should_do_something_for_todays_half/
)
And [See this tweet](https://twitter.com/alex_kroeger/status/1487821077669552132?s=20&t=gzYlmQVEG7k5vw6PEphA3A)

TL;DR: Metamask should display a warning dialog whenever user try to send ERC-20 token to ANY Contract Address directly. Many other wallet software already added this feature e.g. rainbow.me, block wallet. 
The current ["known contract address" solution](https://metamask.zendesk.com/hc/en-us/articles/360020028092-What-is-the-known-contract-address-warning-) is far from enough to protect users. 

Here is my proposed warning text:

> WARNING
> The recipient address you typed in is a Contract address.
> Typically, if you want to give asset to a contract, you should use the dApp of the contract, usually in the form of a website, then follow the instructions there. NOT transferring it here.
> We do not know if the recipient contract can handle a direct token transfer like what you are trying to do, or not. Usually, if it is an DeFi contract, or a token contract, it does not have the ability.
> If that’s the case and you proceed, ALL ASSET SENT will be PERMANENTLY LOST, there is NO WAY to RECOVER.
> Are you sure you want to proceed?
> Cancel (In Big Red Button).
> I want to proceed anyway.
> Option 1: It’s my smart wallet. Mark it as my smart wallet address. Proceed.
> Option 2: I’m very sure this is what I want to do. Proceed.

Full article:
A man just lost 195 WETH, by sending it back to WETH contract address, thinking it would return him raw ETH. He was mislead by introductions on Binance Academy (see original post on r/ethereum). Three more victims showed up today, losing 0.1 0.05 0.03 WETH respectively, check WETH contract addresses’ own WETH holding.
Since ERC-20 standard does not provide a way to notify contracts when they receive token transfers, and it does not mandate the contract the ability to send tokens back out, people who bring the real world common sense may think it wrongly as “I sent it to the account get Wrapped token, now I send it back I should get raw token back”, result in permanent unrecoverable loss.
Vitalik pointed out that Smart, Self-Custodial Wallets are Smart Contracts too, so we cannot ban all transfers to Contract Address all together. I suggest Metamask on all platforms show a warning as shown above when user try to send any ERC-20 token to a Contract Address.
It might be the last chance before unrecoverable loss happens, for many users to realize that it is not the correct operation to do what they want to achieve.

### Steps to reproduce

The current "known contract address" solution at https://metamask.zendesk.com/hc/en-us/articles/360020028092-What-is-the-known-contract-address-warning- is far from enough to protect users. 
This mechanism must detect any contract address by querying ethereum backend, because you can't whitelist all of them.
The contract address of WETH9 token is 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2 . You can all try it out to copy paste it into your Metamask wallet and there will be absolutely ZERO warning about it being a contract address.

### Error messages or log output

_No response_

### Version

10.8.1

### Build type

_No response_

### Browser

Chrome

### Operating system

MacOS

### Hardware wallet

Ledger

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by danjm**
Fixes: #12642
Fixes: #13486
Replaces #12983

This PR ensures we are showing sufficient warnings to users when they are sending to token contracts on their current network or when they are sending to known contract addresses on mainnet.

Screen captures of new behaviour

https://user-images.githubusercontent.com/7499938/153390932-93227c67-a299-4a33-9a7a-e88a8d7d9f6d.mp4

https://user-images.githubusercontent.com/7499938/153390934-2df9b0d2-69d1-4a76-992c-909bad4ab9f7.mp4

### CODE REVIEW

**Review by darkwing — DISMISSED** [review]
This looks good, but needs tests passing.

**Comment by ryanml** [comment]
Now using V2 of `ActionableMessage`:
Screen Shot 2022-02-17 at 11 34 22 AM (1)

**Comment by darkwing** [comment]
From a code perspective this is awesome.  I do have a suggestion about messaging though.  Could we change to something like`Warning: you are about to send to a token contract which could result in a loss of funds. Click to proceed. $1`.  I think we could be a bit more alarmist here.

**Comment by holantonela** [comment]
Thanks for this suggestion @darkwing. Let's move on with that copy @ryanml.

**Comment by ryanml** [comment]
Copy updated

File: ui/pages/send/send-content/send-content.component.js [hunk 1]
```diff
@@ -124,6 +140,36 @@ export default class SendContent extends Component {
...
+    const { t } = this.context;
+    return (
+      <div className="send__warning-container">
+        <ActionableMessage
+          type="danger"
+          useIcon
+          iconFillColor="#d73a49"
+          primaryActionV2={{
+            label: t('tryAnywayOption'),
+            onClick: () => acknowledgeRecipientWarning(),
```
Comment (brad-decker) [inline ui/pages/send/send-content/send-content.component.js:154]: [REQ] just pass method name as it takes no props, the intermediate method is unnecessary.

File: ui/pages/send/send-content/send-content.component.js [hunk 2]
```diff
@@ -124,6 +140,36 @@ export default class SendContent extends Component {
...
+          }}
+          message={t('sendingToTokenContractWarning', [
+            <a
+              key="contractWarningSupport"
+              target="_blank"
+              rel="noopener noreferrer"
+              className="send__warning-container__link"
+              href="https://metamask.zendesk.com/hc/en-us/articles/360020028092-What-is-the-known-contract-address-warning-"
```
Comment (brad-decker) [inline ui/pages/send/send-content/send-content.component.js:162]: [REQ] There's a new zendesk urls constant in the constants folder we should use for this.

File: ui/ducks/send/send.js [hunk 3]
```diff
@@ -1061,52 +1068,57 @@ const slice = createSlice({
       // input field
       state.recipient.userInput = action.payload;
     },
-    validateRecipientUserInput: (state, action) => {
+    updateRecipientWarning: (state, action) => {
+      state.recipient.warning = action.payload;
+    },
+    validateRecipientAddress: (state, action) => {
```
Comment (brad-decker) [inline ui/ducks/send/send.js:1074]: [Q] I'm not against it just curious why this was changed? its been a while but i thought that the original intent was to validate the user's input *Before setting it as the actual address of the recipient in state*

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: ui/pages/send/send-content/send-content.component.js (hunk 1 above)
Comment (ryanml) [reply ui/pages/send/send-content/send-content.component.js]: Fixed

File: ui/pages/send/send-content/send-content.component.js (hunk 2 above)
Comment (ryanml) [reply ui/pages/send/send-content/send-content.component.js]: done

File: ui/ducks/send/send.js (hunk 3 above)
Comment (ryanml) [reply ui/ducks/send/send.js:1074]: cc: @danjm

**Review by brad-decker — DISMISSED** [review]

File: ui/ducks/send/send.js [hunk 4]
```diff
@@ -1061,52 +1068,57 @@ const slice = createSlice({
...
       const { asset, recipient } = state;
 
       if (
         recipient.mode === RECIPIENT_SEARCH_MODES.MY_ACCOUNTS ||
-        recipient.userInput === '' ||
-        recipient.userInput === null
+        recipient.address === '' ||
```
Comment (Gtonizuka) [inline ui/ducks/send/send.js:1079]: can this ever be undefined? If so we could just get away with `if(Boolean(recipient.address )`

File: ui/ducks/send/send.js [hunk 5]
```diff
@@ -1530,31 +1552,38 @@ export function updateSendAsset({ type, details }) {
...
     const state = getState();
     const chainId = getCurrentChainId(state);
     const tokens = getTokens(state);
     const useTokenDetection = getUseTokenDetection(state);
     const tokenAddressList = Object.keys(getTokenList(state));
-    debouncedValidateRecipientUserInput(dispatch, {
+    const { symbol, decimals } = await getSymbolAndDecimals(
+      userInput,
+      tokenAddressList,
+    );
```
Comment (adonesky1) [inline ui/ducks/send/send.js:1579]: We might consider using `[getTokenStandardAndDetails](https://github.com/MetaMask/controllers/blob/a799f42ad7448a951ae02fd00be7fbfe6904aa8e/src/assets/AssetsContractController.ts#L167)` since it actually returns the contract standard and a more details with which we can inform the user more clearly about the contract they are sending to.

Comment (adonesky1) [reply ui/ducks/send/send.js:1579]: ☝️  This would also allow us to tell users if they are sending to an ERC721 or ERC1155 contract, which this current approach doesn't enable.

File: ui/ducks/send/send.js (hunk 3 above)
Comment (adonesky1) [reply ui/ducks/send/send.js:1074]: ➕

Comment (adonesky1) [reply ui/ducks/send/send.js:1074]: It looks like you're setting recipient.address to `''` right before calling into here and then checking whether its `''`
on line [1642](https://github.com/MetaMask/metamask-extension/pull/13588/files#diff-bad76d7279418f3520a0ca93cfaeeb6fcbb1f0d4782f6ff942d422632a1c2780L1611)

Comment (adonesky1) [reply ui/ducks/send/send.js:1074]: oh wait nvm thats in a reset method

**Review by DELJUJAZI — DISMISSED** [review]

File: ui/ducks/send/send.js (hunk 3 above)
Comment (darkwing) [reply ui/ducks/send/send.js:1118]: This appears to still require an answer.

File: ui/ducks/send/send.js (hunk 4 above)
Comment (darkwing) [reply ui/ducks/send/send.js:1123]: I like this idea!

File: ui/ducks/send/send.js [hunk 6]
```diff
@@ -1313,10 +1323,14 @@ const slice = createSlice({
         recipient.error = null;
         recipient.warning = null;
       } else {
```
Comment (danjm) [inline ui/ducks/send/send.js:1325]: `isSendingToken` is removed because we want to show the user a warning regardless of what currency they are sending to a token contract

File: ui/ducks/send/send.js [hunk 7]
```diff
@@ -1821,14 +1844,18 @@ export function updateSendAsset({ type, details }) {
...
  */
-const debouncedValidateRecipientUserInput = debounce((dispatch, payload) => {
-  dispatch(
-    addHistoryEntry(
-      `sendFlow - user typed ${payload.userInput} into recipient input field`,
-    ),
-  );
-  dispatch(validateRecipientUserInput(payload));
-}, 300);
+const debouncedValidateRecipientUserInput = debounce(
```
Comment (danjm) [inline ui/ducks/send/send.js:1847]: The addition of resolve as a parameter allows the caller to asynchronously await the debounced call

File: ui/pages/send/send.js [hunk 8]
```diff
@@ -111,7 +112,6 @@ export default function SendTransactionScreen() {
               `sendFlow - User pasted ${text} into address field`,
             ),
           );
```
Comment (danjm) [inline ui/pages/send/send.js:114]: This update recipient call is being removed because pastes also trigger the change listener on the input in the EnsInput component, which results in `onValidAddressTyped` being called, which (as you can see above) also calls `updateRecipient`. We don't want to call that twice.

File: ui/ducks/send/send.js [hunk 9]
```diff
@@ -2486,6 +2559,11 @@ export function getRecipientUserInput(state) {
   return state[name].recipientInput;
 }
 
+export function getRecipientWarningAcknowledgement(state) {
+  return getCurrentDraftTransaction(state).recipient
+    .recipientWarningAcknowledged;
```
Comment (brad-decker) [inline ui/ducks/send/send.js:2564]: Should probably use optional chaining with a sensible default (false?) because the draftTransaction *could be an empty object* if no draft transaction is currently found. This is an edge case but its better to guard against it.

**Review by brad-decker — CHANGES_REQUESTED** [review]

**Comment by darkwing** [comment]
I was able to trigger a weird error with this:

STR:
1.  Click "Send" from the home screen
2.  Use the Chainlink contract address:  `0x514910771AF9Ca656af840dff83E8264EcF986CA`
3.  See the contract warning message (yay!)
4.  Click the currency box
5. See that the contract warning goes away, and the screen shows a "new address" block


https://user-images.githubusercontent.com/46655/178345905-79cfd968-d4bd-4112-a8ef-3ed50e8c14fc.mp4



I was also able to trigger an actual error by clicking on something else, but don't recall how:

WeirdError2

**Comment by danjm** [comment]
> Click "Send" from the home screen
> Use the Chainlink contract address: 0x514910771AF9Ca656af840dff83E8264EcF986CA
> See the contract warning message (yay!)
> Click the currency box
> See that the contract warning goes away, and the screen shows a "new address" block

@darkwing That's not a bug. We always show that "new address" block when sending to a new address. We could claim that the UX of showing one block after another is not good, but I don't think that is a bug.

Meanwhile, the type error identified is a bug on develop: https://github.com/MetaMask/metamask-extension/issues/15188

File: ui/pages/send/send.js [hunk 10]
```diff
@@ -96,10 +96,11 @@ export default function SendTransactionScreen() {
         userInput={userInput}
         className="send__to-row"
         onChange={(address) => dispatch(updateRecipientUserInput(address))}
-        onValidAddressTyped={(address) => {
+        onValidAddressTyped={async (address) => {
           dispatch(
             addHistoryEntry(`sendFlow - Valid address typed ${address}`),
           );
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/pages/send/send.js:97]: @brad-decker Can you add more clarification/detail to this. `updateRecipient` is not called in the debounce handler called from `updateRecipientUserInput` so I'm not exactly sure what you are suggesting.

File: ui/ducks/send/send.js (hunk 9 above)
Comment (danjm) [reply ui/ducks/send/send.js:2571]: done in ca614adcb

**Comment by danjm** [comment]
> That's not a bug. We always show that "new address" block when sending to a new address. We could claim that the UX of showing one block after another is not good, but I don't think that is a bug.

I was wrong about that, I misunderstood the bug.

The bug was that the recipient warning went away went clicking the Asset dropdown. I fixed this with 89428b9ab

**Review by ryanml — APPROVED** [review]

**Review by brad-decker — APPROVED** [review]
